src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java

```diff
@@ -312,4 +340,27 @@ public String getDisplayName() {
             return "Parallel test job execution";
         }
     }
+
+    public static class StageNodeFinderPredicate implements Predicate<FlowNode> {
+        private final String stageName;
+
+        public StageNodeFinderPredicate(@Nonnull String stageName) {
+            this.stageName = stageName;
+        }
+
+        @Override
+        public boolean apply(@CheckForNull FlowNode input) {
+            if (input instanceof BlockStartNode &&
+                    input instanceof StepNode) {
+                StepDescriptor descriptor = ((StepNode)input).getDescriptor();
+                LabelAction label = input.getPersistentAction(LabelAction.class);
+
+                return descriptor != null &&
```

Comment:
🐜 This whole thing can be replaced with:

```java
static final predicate STAGE_PREDICATE = Predicates.and(new NodeStepTypePredicate("stage"), FlowScanningUtils.hasActionPredicate(LabelAction.class))
```

Reply:
Not *quite* - I also need to check the stage name. But lemme see what I can do.

Reply:
Ah, yeah -- looks like I need that second cup of coffee today before doing reviews.   But in any case you want to check for the descriptor *before* doing the action check (that can be done with the existing predicate).  98% of the time the descriptor won't match, and that check is cheaper than actions (even after optimization) and the other 2% are what we care about.   If we had a static method on LabelAction `@CheckForNull static String getLabel(FlowNode)` this might be simplified a bit.